File: python/tests/comms/mcp2515_test.py
from machine import SPI, Pin
from periph.connection.spi_micropython import SPIConnection
from periph.chips.comms.mcp2515 import MCP2515Minimal, MCP2515Full, CanFrame

passed = 0
failed = 0

def check_eq(label, got, expected):
    global passed, failed
    if got == expected:
        print('PASS', label); passed += 1
    else:
        print('FAIL {}: got {!r}, expected {!r}'.format(label, got, expected)); failed += 1

def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label); passed += 1
    else:
        print('FAIL', label); failed += 1


spi = SPI(1, baudrate=10_000_000, polarity=0, phase=0)
cs = Pin(5, Pin.OUT)
connection = SPIConnection(spi, cs)
can = MCP2515Minimal(connection)

# CanFrame validation
try:
    CanFrame(0x800, b'', extended=False)
    check_true('CanFrame rejects standard id > 0x7FF', False)
except ValueError:
    check_true('CanFrame rejects standard id > 0x7FF', True)

try:
    CanFrame(0x20000000, b'', extended=True)
    check_true('CanFrame rejects extended id > 0x1FFFFFFF', False)
except ValueError:
    check_true('CanFrame rejects extended id > 0x1FFFFFFF', True)

try:
    CanFrame(0x123, b'\x00' * 9)
    check_true('CanFrame rejects dlc > 8', False)
except ValueError:
    check_true('CanFrame rejects dlc > 8', True)

# Send a standard 11-bit frame
can.send(0x123, b'\xde\xad\xbe\xef', extended=False)
check_eq('send standard id',  can._read_status() is not None, True)
check_eq('send returns None', None, None)

# Construct Full class on the same connection and exercise Full-only API
can_full = MCP2515Full(connection)
can_full.set_mode('loopback')
check_eq('set_mode loopback', can_full.get_mode(), 'loopback')

can_full.set_mode('normal')
check_eq('set_mode normal', can_full.get_mode(), 'normal')

can_full.set_one_shot(True)
check_eq('set_one_shot enable', can_full._read_reg(0x0F) & 0x08, 0x08)

can_full.set_one_shot(False)
check_eq('set_one_shot disable', can_full._read_reg(0x0F) & 0x08, 0x00)

errs = can_full.read_errors()
check_true('read_errors returns dict', isinstance(errs, dict))
check_true('read_errors has tec', 'tec' in errs)
check_true('read_errors has rec', 'rec' in errs)
check_true('read_errors has eflg', 'eflg' in errs)

print('===DONE: {} passed, {} failed==='.format(passed, failed))

File: nodejs/tests/comms/mcp2515_test.js
'use strict';

const { SPIConnection } = require('../../packages/periph/src/connection/spi');
const { MCP2515Full }    = require('../../packages/periph/src/chips/comms/mcp2515');

const SPI_BUS = parseInt(process.env.SPI_BUS  || '0', 10);
const SPI_DEV = parseInt(process.env.SPI_DEV  || '0', 10);
const BITRATE = parseInt(process.env.MCP2515_BITRATE || '125', 10);

let passed = 0;
let failed = 0;

function checkEq(label, got, expected) {
    if (got === expected) { console.log('PASS', label); passed++; }
    else { console.log(`FAIL ${label}: got ${got}, expected ${expected}`); failed++; }
}

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else { console.log(`FAIL ${label}`); failed++; }
}

(async () => {
    const connection = new SPIConnection(SPI_BUS, SPI_DEV, { mode: 0, maxSpeedHz: 10_000_000 });
    const chip = new MCP2515Full(connection, BITRATE);

    checkEq('init_mode', await chip.getMode(), 'normal');

    await chip.setMode('loopback');
    checkEq('loopback_mode', await chip.getMode(), 'loopback');

    await chip.setMode('normal');
    checkEq('normal_mode_after_loopback', await chip.getMode(), 'normal');

    await chip.send(0x7FF, Buffer.from([0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08]), false);
    const frame1 = await chip.recv(100);
    checkTrue('recv_standard', frame1 !== null && frame1.id === 0x7FF);

    await chip.send(0x1FFFFFFF, Buffer.from([0xAA, 0xBB, 0xCC, 0xDD]), true);
    const frame2 = await chip.recv(100);
    checkTrue('recv_extended', frame2 !== null && frame2.id === 0x1FFFFFFF && frame2.extended === true);

    const errors = await chip.readErrors();
    checkTrue('read_errors', typeof errors.tec === 'number' && typeof errors.rec === 'number' && typeof errors.eflg === 'number');

    await chip.setOneShot(true);
    await chip.setOneShot(false);

    await chip.setFilter(0, 0x100, false);
    await chip.setMask(0, 0x7FF, false);
    await chip.setRxMode(0, 0);

    await chip.reset();
    await chip.init(BITRATE, 8);
    checkEq('init_after_reset', await chip.getMode(), 'normal');

    await connection.close();

    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
})();
